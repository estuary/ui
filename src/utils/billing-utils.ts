import { isEqual, parseISO } from 'date-fns';
import { isEmpty, sum } from 'lodash';
import prettyBytes from 'pretty-bytes';
import { BillingRecord } from 'stores/Billing/types';
import { CatalogStats_Billing, Entity } from 'types';

export const TOTAL_CARD_HEIGHT = 300;

// Grid item height - 72 = graph canvas height
export const CARD_AREA_HEIGHT = TOTAL_CARD_HEIGHT - 72;

export const BYTES_PER_GB = 1073741824;

const FREE_TASK_COUNT = 2;

export enum FREE_GB_BY_TIER {
    FREE = 20,
    PERSONAL = 10,
    ENTERPRISE = 10,
}

export const evaluateSpecType = (query: CatalogStats_Billing): Entity => {
    if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
        return 'capture';
    } else if (Object.hasOwn(query.flow_document.taskStats, 'materialize')) {
        return 'materialization';
    } else {
        return 'collection';
    }
};

const evaluateTotalCost = (dataVolume: number, taskCount: number) => {
    const freeBytes = 10 * BYTES_PER_GB;

    const dataVolumeOverLimit =
        dataVolume > freeBytes ? dataVolume - freeBytes : 0;

    const taskCountOverLimit =
        taskCount > FREE_TASK_COUNT ? taskCount - FREE_TASK_COUNT : 0;

    return (
        (dataVolumeOverLimit / BYTES_PER_GB) * 0.75 + taskCountOverLimit * 20
    );
};

const evaluateDataVolume = (stats: CatalogStats_Billing[]): number => {
    const taskBytes: number[] = stats.map((query) => {
        if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
            return query.bytes_written_by_me;
        } else if (
            Object.hasOwn(query.flow_document.taskStats, 'materialize')
        ) {
            return query.bytes_read_by_me;
        } else {
            return query.bytes_written_by_me + query.bytes_read_by_me;
        }
    });

    return sum(taskBytes);
};

export const stripTimeFromDate = (date: string) => {
    const [truncatedDateStr] = date.split('T', 1);

    return parseISO(truncatedDateStr);
};

const getInitialBillingRecord = (date: string): BillingRecord => {
    const truncatedDate = stripTimeFromDate(date);

    return {
        date: truncatedDate,
        dataVolume: 0,
        taskCount: 0,
        totalCost: 0,
        pricingTier: null,
        taskRate: null,
        gbFree: null,
        includedTasks: null,
    };
};

// TODO (billing): Remove this helper function to translate data returned from
//   the new RPC when available.
export const formatBillingCatalogStats = (
    value: CatalogStats_Billing[]
): BillingRecord[] => {
    const taskStatData = value.filter((query) =>
        Object.hasOwn(query.flow_document, 'taskStats')
    );

    let sortedStats: {
        [date: string]: CatalogStats_Billing[];
    } = {};

    taskStatData.forEach((query) => {
        if (Object.hasOwn(sortedStats, query.ts)) {
            sortedStats[query.ts].push(query);
        } else {
            sortedStats = {
                ...sortedStats,
                [query.ts]: [query],
            };
        }
    });

    const billingHistory: BillingRecord[] = [];

    if (!isEmpty(sortedStats)) {
        Object.entries(sortedStats).forEach(([ts, stats]) => {
            const billingRecordIndex = billingHistory.findIndex((record) =>
                isEqual(record.date, stripTimeFromDate(ts))
            );

            const taskCount = stats.length;
            const dataVolume = evaluateDataVolume(stats);
            const totalCost = evaluateTotalCost(dataVolume, taskCount);

            if (billingRecordIndex === -1) {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    getInitialBillingRecord(ts);

                billingHistory.push({
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                });
            } else {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    billingHistory[billingRecordIndex];

                billingHistory[billingRecordIndex] = {
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                };
            }
        });
    }

    return billingHistory;
};

export interface SeriesConfig {
    data: [string, number][];
    seriesName?: string;
    stack?: string;
}

export const formatDataVolumeForDisplay = (
    seriesConfigs: SeriesConfig[],
    tooltipConfig: any
): string => {
    const filteredSeries =
        seriesConfigs.length === 1
            ? seriesConfigs
            : seriesConfigs.filter(
                  (series) => series.seriesName === tooltipConfig.seriesName
              );

    const dataVolumeInBytes = filteredSeries
        .flatMap(({ data }) => data)
        .find(([month]) => month === tooltipConfig.name);

    return dataVolumeInBytes
        ? prettyBytes(dataVolumeInBytes[1], { minimumFractionDigits: 2 })
        : `${tooltipConfig.value[1]} GB`;
};
