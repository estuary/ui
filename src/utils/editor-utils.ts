export const ICON_SIZE = 14;

export const DEFAULT_TOOLBAR_HEIGHT = 20;
export const DEFAULT_HEIGHT = 330;

// The two pixels added to the composite height accounts for
// the presence of a top and bottom, pixel-wide border.
export const getEditorTotalHeight = (
    editorHeight?: number,
    toolbarHeight?: number
) => {
    return (
        (editorHeight ?? DEFAULT_HEIGHT) +
        (toolbarHeight ?? DEFAULT_TOOLBAR_HEIGHT) +
        2
    );
};
