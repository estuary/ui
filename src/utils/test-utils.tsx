/* eslint-disable @typescript-eslint/unbound-method */
// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { AuthContext } from 'context/Auth';
import AppContent from 'context/Content';
import AppRouter from 'context/Router';
import AppTheme from 'context/Theme';
import { ReactElement } from 'react';
import { auth, AuthDetails } from 'services/auth';

// TODO - this does not really do anything because we hardcode
//   the username down below on the MockAuthProvider. However,
//   this will be needed in the future so leaving the work so far
//   so we can leverage this when wanting to test the "on load auth"
//   functionality
const loginAsUser = (username: string) => {
    const mockAuthDetails: AuthDetails = {
        session: {
            account_id: '',
            expires_at: '',
            token: '',
        },
        user: {
            created_at: '',
            display_name: username,
            email: `${username}@${username}`,
            id: '',
            name: username,
            unique_name: username,
            updated_at: '',
        },
    };

    auth.saveSession(mockAuthDetails.session);
    auth.saveUser(mockAuthDetails.user);
};

const logoutUser = () => {
    auth.removeAuthDetails();
};

const goTo = (route?: string, name?: string) => {
    window.history.pushState(
        {},
        name ? name : 'Test page',
        route ? route : '/'
    );
};

const waitForLoading = async () => {
    return Promise.resolve();
};

// Mocking AuthContext out basically skips all the loading screen and boostrapping
//  stuff that happens on load. This should be fine for unit testing as it
//  can help speed up tests a bit and makes things easier to deal with. This
//  does mean that we need to eventually write a basic test that handles the
//  account bootstrapping.
interface MockAuthProps {
    children: JSX.Element;
    username?: string;
}
const MockAuthProvider = ({ children, username }: MockAuthProps) => {
    const value = {
        login: jest.fn(() => {
            if (username) {
                loginAsUser(username);
            }
            return Promise.resolve();
        }),
        logout: jest.fn(() => {
            logoutUser();
            return Promise.resolve();
        }),
        user: username,
    };

    return (
        <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
};

const customRender = async (
    ui: ReactElement,
    options: Omit<RenderOptions, 'wrapper'> & {
        route?: string;
        user?: string;
    }
) => {
    const { route, user } = options;

    goTo(route, 'Test Page');

    const view = rtlRender(
        <AppContent>
            <MockAuthProvider username={user}>
                <AppTheme>
                    <AppRouter>{ui}</AppRouter>
                </AppTheme>
            </MockAuthProvider>
        </AppContent>,
        {
            ...options,
        }
    );

    await waitForLoading();

    return view;
};

export * from '@testing-library/react';
export { goTo, loginAsUser, logoutUser, customRender, waitForLoading };
