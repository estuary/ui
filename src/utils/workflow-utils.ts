import type { ConnectorConfig } from 'deps/flow/flow';
import type { DraftSpecsExtQuery_ByCatalogName } from 'src/api/draftSpecs';
import type { DraftSpecQuery } from 'src/hooks/useDraftSpecs';
import type { CallSupabaseResponse } from 'src/services/supabase';
import type {
    FullSource,
    FullSourceDictionary,
} from 'src/stores/Binding/slices/TimeTravel';
import type {
    Bindings,
    ResourceConfigDictionary,
} from 'src/stores/Binding/types';
import type {
    DekafConfig,
    Entity,
    EntityWithCreateWorkflow,
    Schema,
    SourceCaptureDef,
} from 'src/types';
import type {
    BuiltBinding,
    MaterializationBinding,
    ValidatedBinding,
} from 'src/types/schemaModels';

import { isBoolean, isEmpty, isEqual } from 'lodash';

import { modifyDraftSpec } from 'src/api/draftSpecs';
import { isDekafEndpointConfig } from 'src/utils/connector-utils';
import {
    addOrRemoveOnIncompatibleSchemaChange,
    addOrRemoveSourceCapture,
} from 'src/utils/entity-utils';
import { hasLength } from 'src/utils/misc-utils';

// This is the soft limit we recommend to users
export const MAX_BINDINGS = 300;
export const DEFAULT_DEBOUNCE_WAIT = 500;
export const QUICK_DEBOUNCE_WAIT = DEFAULT_DEBOUNCE_WAIT / 2;
export const NEAR_INSTANT_DEBOUNCE_WAIT = QUICK_DEBOUNCE_WAIT / 2;
export const NEW_TASK_PUBLICATION_ID = '00:00:00:00:00:00:00:00';

export const getBackfillCounter = (binding: any): number => {
    return Object.hasOwn(binding, 'backfill') ? binding.backfill : 0;
};

export const getSourceOrTarget = (binding: any) => {
    return Object.hasOwn(binding ?? {}, 'source')
        ? binding.source
        : Object.hasOwn(binding ?? {}, 'target')
          ? binding.target
          : binding;
};

export const getBindingAsFullSource = (binding: any) => {
    const response = getSourceOrTarget(binding);
    if (typeof response === 'string') {
        return {
            name: response,
        };
    }

    return getSourceOrTarget(binding);
};

export const getCollectionNameProp = (entityType: Entity) => {
    return entityType === 'materialization' ? 'source' : 'target';
};

export const getCollectionNameDirectly = (binding: any) => {
    // Check if we're dealing with a FullSource or just a string
    return Object.hasOwn(binding ?? {}, 'name') ? binding.name : binding;
};

export const getCollectionName = (binding: any) => {
    // First see if we've already been passed a scoped binding
    //  or if we need to find the proper scope ourselves.
    const scopedBinding = getSourceOrTarget(binding);

    return getCollectionNameDirectly(scopedBinding);
};

export const isValidIndex = (index: number) => index > -1;

export const getBindingIndex = (
    existingBindings: any[],
    collectionName: string,
    stagedBindingIndex: number
) => {
    const matchedCollectionIndices = existingBindings
        .map((binding: any, index) => {
            if (getCollectionName(binding) === collectionName) {
                return index;
            } else {
                return -1;
            }
        })
        .filter(isValidIndex);

    if (matchedCollectionIndices.length === 1) {
        return matchedCollectionIndices[0];
    } else if (
        stagedBindingIndex > -1 &&
        matchedCollectionIndices.length > stagedBindingIndex
    ) {
        return matchedCollectionIndices[stagedBindingIndex];
    } else {
        return -1;
    }
};

export const getDisableProps = (disable: boolean | undefined) => {
    return disable ? { disable } : {};
};

// Used to mark fields that should be removed during generation. This is
//      only here because if we set something to null and then check for nulls
//      we might end up overwritting a value a user specifically wants a null for.
export const REMOVE_DURING_GENERATION = undefined;
export const getFullSource = (
    fullSource: FullSource | string | undefined,
    filterOutName?: boolean,
    filterOutRemovable?: boolean
): {
    fullSource?: FullSource;
} => {
    if (typeof fullSource === 'string' || fullSource === undefined) {
        return {};
    }

    const response = {
        fullSource: { ...fullSource },
    };

    if (filterOutName) {
        delete response.fullSource.name;
    }

    if (filterOutRemovable) {
        response.fullSource = Object.entries(response.fullSource).reduce(
            (filtered, [key, val]) => {
                if (val !== REMOVE_DURING_GENERATION) {
                    filtered[key] = val;
                }

                return filtered;
            },
            {}
        );
    }

    return response;
};

export const getFullSourceSetting = (
    fullSource: FullSourceDictionary | null,
    collectionName: string,
    bindingUUID: string
) => {
    const fullSourceConfig = fullSource?.[bindingUUID]?.data;

    return !isEmpty(fullSourceConfig)
        ? { ...fullSourceConfig, name: collectionName }
        : collectionName;
};

export const updateFullSource = () => {};

// TODO (typing): Narrow the return type for this function.
export const generateTaskSpec = (
    entityType: EntityWithCreateWorkflow,
    endpointConfig: ConnectorConfig | DekafConfig,
    resourceConfigs: ResourceConfigDictionary,
    resourceConfigServerUpdateRequired: boolean,
    bindings: Bindings,
    existingTaskData: DraftSpecsExtQuery_ByCatalogName | null,
    options: {
        fullSource: FullSourceDictionary | null;
        sourceCaptureDefinition: SourceCaptureDef | null;
        specOnIncompatibleSchemaChange?: string;
    }
) => {
    const draftSpec = isEmpty(existingTaskData)
        ? {
              bindings: [],
              endpoint: {},
          }
        : existingTaskData.spec;

    const endpointProp = isDekafEndpointConfig(endpointConfig)
        ? 'dekaf'
        : 'connector';

    draftSpec.endpoint = { [endpointProp]: endpointConfig };

    if (!isEmpty(resourceConfigs) && !isEmpty(bindings)) {
        const collectionNameProp = getCollectionNameProp(entityType);
        const { fullSource } = options;

        Object.entries(bindings).forEach(([_collection, bindingUUIDs]) => {
            bindingUUIDs.forEach((bindingUUID, iteratedIndex) => {
                const resourceConfig = resourceConfigs[bindingUUID].data;

                const {
                    bindingIndex,
                    collectionName,
                    disable,
                    onIncompatibleSchemaChange,
                } = resourceConfigs[bindingUUID].meta;

                // Check if disable is a boolean otherwise default to false
                const bindingDisabled = isBoolean(disable) ? disable : false;

                // See which binding we need to update
                const existingBindingIndex = resourceConfigServerUpdateRequired
                    ? getBindingIndex(
                          draftSpec.bindings,
                          collectionName,
                          iteratedIndex
                      )
                    : hasLength(draftSpec.bindings)
                      ? bindingIndex
                      : -1;

                if (existingBindingIndex > -1) {
                    // Include disable otherwise totally remove it
                    if (bindingDisabled) {
                        draftSpec.bindings[existingBindingIndex].disable =
                            bindingDisabled;
                    } else {
                        delete draftSpec.bindings[existingBindingIndex].disable;
                    }

                    if (entityType === 'materialization') {
                        addOrRemoveOnIncompatibleSchemaChange(
                            draftSpec.bindings[existingBindingIndex],
                            onIncompatibleSchemaChange
                        );
                    }

                    // Only update if there is a fullSource to populate. Otherwise just set the name.
                    //  This handles both captures that do not have these settings AND when
                    draftSpec.bindings[existingBindingIndex] = {
                        ...draftSpec.bindings[existingBindingIndex],
                        resource: {
                            ...resourceConfig,
                        },
                        [collectionNameProp]: getFullSourceSetting(
                            fullSource,
                            collectionName,
                            bindingUUID
                        ),
                    };
                } else if (Object.keys(resourceConfig).length > 0) {
                    const disabledProps = getDisableProps(bindingDisabled);

                    const newBinding: Schema = {
                        [collectionNameProp]: getFullSourceSetting(
                            fullSource,
                            collectionName,
                            bindingUUID
                        ),
                        ...disabledProps,
                        resource: {
                            ...resourceConfig,
                        },
                    };

                    if (entityType === 'materialization') {
                        addOrRemoveOnIncompatibleSchemaChange(
                            newBinding,
                            onIncompatibleSchemaChange
                        );
                    }

                    draftSpec.bindings.push(newBinding);
                }
            });
        });

        if (hasLength(draftSpec.bindings)) {
            const boundCollections = Object.keys(bindings);

            draftSpec.bindings = draftSpec.bindings.filter((binding: any) =>
                boundCollections.includes(
                    getCollectionName(binding[collectionNameProp])
                )
            );
        }
    } else {
        draftSpec.bindings = [];
    }

    // Try adding at the end because these settings could be added/changed at any time
    if (entityType === 'materialization') {
        addOrRemoveSourceCapture(draftSpec, options.sourceCaptureDefinition);
        addOrRemoveOnIncompatibleSchemaChange(
            draftSpec,
            options.specOnIncompatibleSchemaChange
        );
    }

    return draftSpec;
};

// const mergeResourceConfigs = (
//     queryData: DraftSpecQuery,
//     resourceConfig: ResourceConfigDictionary,
//     restrictedDiscoveredCollections: string[]
// ): ResourceConfigDictionary => {
//     const existingCollections = Object.keys(resourceConfig);
//     const mergedResourceConfig: ResourceConfigDictionary = {};

//     Object.entries(resourceConfig).forEach(([key, value]) => {
//         mergedResourceConfig[key] = value;
//     });

//     queryData.spec.bindings.forEach((binding: any) => {
//         if (
//             !existingCollections.includes(binding.target) &&
//             !restrictedDiscoveredCollections.includes(binding.target)
//         ) {
//             mergedResourceConfig[binding.target] = {
//                 data: binding.resource,
//                 errors: [],
//             };
//         }
//     });

//     return mergedResourceConfig;
// };

export interface SupabaseConfig {
    lastPubId: string;
    catalogName?: string;
}

export const modifyDiscoveredDraftSpec = async (
    response: {
        data: DraftSpecQuery[];
        error?: undefined;
    },
    supabaseConfig?: SupabaseConfig | null
): Promise<CallSupabaseResponse<any>> => {
    const draftSpecData = response.data[0];

    return modifyDraftSpec(
        draftSpecData.spec,
        {
            draft_id: draftSpecData.draft_id,
            catalog_name: draftSpecData.catalog_name,
        },
        supabaseConfig?.catalogName,
        supabaseConfig?.lastPubId
    );
};

export const modifyExistingCaptureDraftSpec = async (
    draftId: string,
    connectorImage: string,
    encryptedEndpointConfig: Schema,
    resourceConfig: ResourceConfigDictionary,
    existingTaskData: DraftSpecsExtQuery_ByCatalogName | null,
    resourceConfigServerUpdateRequired: boolean,
    bindings: Bindings
): Promise<CallSupabaseResponse<any>> => {
    const draftSpec = generateTaskSpec(
        'capture',
        { image: connectorImage, config: encryptedEndpointConfig },
        resourceConfig,
        resourceConfigServerUpdateRequired,
        bindings,
        existingTaskData,
        {
            fullSource: null,
            sourceCaptureDefinition: null,
        }
    );

    return modifyDraftSpec(draftSpec, {
        draft_id: draftId,
        spec_type: 'capture',
    });
};

export const getBuiltBindingIndex = (
    builtSpec: Schema,
    targetCollection: string
): number => {
    const builtBindings: BuiltBinding[] = builtSpec.bindings ?? [];

    return builtBindings.findIndex(
        (binding) => binding.collection.name === targetCollection
    );
};

export const getBindingIndexByResourcePath = <T extends Schema>(
    resourcePath: string[],
    schema: Schema
): number => {
    if (resourcePath.length === 0) {
        return -1;
    }

    const bindings: T[] = schema.bindings;

    return bindings.findIndex((binding) => {
        let bindingResourcePath: string[] = [];

        if (binding?.resourcePath) {
            bindingResourcePath = binding.resourcePath;
        } else if (binding?.resource?._meta?.path) {
            bindingResourcePath = binding.resource._meta.path;
        }

        return bindingResourcePath.length > 0
            ? isEqual(bindingResourcePath, resourcePath)
            : false;
    });
};

const getDraftedMaterializationBinding = (
    draftSpec: Schema,
    resourcePath: string[] | undefined,
    stagedBindingIndex: number,
    targetCollection: string,
    indexLookupOnly?: boolean
): {
    draftedBinding: MaterializationBinding | undefined;
    draftedBindingIndex: number;
} => {
    let draftedBinding: MaterializationBinding | undefined;
    let draftedBindingIndex = -1;

    // Attempt to identify the target drafted binding by resource path pointer(s) before
    // relying on the binding index.
    if (!indexLookupOnly && resourcePath) {
        draftedBindingIndex =
            getBindingIndexByResourcePath<MaterializationBinding>(
                resourcePath,
                draftSpec
            );

        draftedBinding = draftSpec.bindings.at(draftedBindingIndex);
    }

    if (!draftedBinding) {
        draftedBindingIndex = getBindingIndex(
            draftSpec.bindings,
            targetCollection,
            stagedBindingIndex
        );

        draftedBinding =
            draftedBindingIndex > -1
                ? draftSpec.bindings[draftedBindingIndex]
                : undefined;
    }

    return { draftedBinding, draftedBindingIndex };
};

interface BindingTypes<B, V, T> {
    built: B | undefined;
    drafted: T | undefined;
    live: T | undefined;
    validated: V | undefined;
}

export interface RelatedBindings {
    indices: BindingTypes<number, number, number>;
    values: BindingTypes<
        BuiltBinding,
        ValidatedBinding,
        MaterializationBinding
    >;
}

export const getRelatedBindings = (
    builtSpec: Schema,
    draftSpec: Schema,
    stagedBindingIndex: number,
    targetCollection: string,
    validationResponse: Schema,
    liveSpec?: Schema | null
): RelatedBindings => {
    // Select the binding from the built spec that corresponds to the current collection
    //  to extract the projection information.
    // Defaulting to empty array. This is to handle when a user has disabled a collection
    //  which causes the binding to not be included in the built_spec
    const builtBindingIndex = getBuiltBindingIndex(builtSpec, targetCollection);
    const builtBinding: BuiltBinding | undefined =
        builtSpec.bindings.at(builtBindingIndex);

    // The validation phase of a publication produces a document which correlates each binding projection
    // to a constraint type (defined in flow/go/protocols/materialize/materialize.proto). Select the binding
    // from the validation document that corresponds to the current collection to extract the constraint types.
    const validatedBindingIndex =
        getBindingIndexByResourcePath<ValidatedBinding>(
            builtBinding?.resourcePath ?? [],
            validationResponse
        );
    const validatedBinding: ValidatedBinding | undefined =
        validationResponse.bindings.at(validatedBindingIndex);

    // TODO (field-selection): Use the staged binding index to identify the target drafted binding for now.
    //   Determine whether resource path pointer can be used for drafted binding lookup.
    const { draftedBinding, draftedBindingIndex } =
        getDraftedMaterializationBinding(
            draftSpec,
            builtBinding?.resourcePath,
            stagedBindingIndex,
            targetCollection,
            true
        );

    const liveBindingIndex = liveSpec
        ? getBindingIndexByResourcePath<MaterializationBinding>(
              builtBinding?.resourcePath ?? [],
              liveSpec
          )
        : undefined;
    const liveBinding: MaterializationBinding | undefined =
        liveSpec?.bindings.at(liveBindingIndex);

    return {
        indices: {
            built: builtBindingIndex,
            drafted: draftedBindingIndex,
            live: liveBindingIndex,
            validated: validatedBindingIndex,
        },
        values: {
            built: builtBinding,
            drafted: draftedBinding,
            live: liveBinding,
            validated: validatedBinding,
        },
    };
};
