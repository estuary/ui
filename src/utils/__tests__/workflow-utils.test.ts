import { DraftSpecsExtQuery_ByCatalogName } from 'api/draftSpecs';
import { FullSourceDictionary } from 'components/editor/Bindings/Store/types';
import {
    ResourceConfig,
    ResourceConfigDictionary,
} from 'stores/ResourceConfig/types';
import { generateMockConnectorConfig } from 'test/test-utils';
import { generateTaskSpec, getBindingIndex } from 'utils/workflow-utils';
import { ConnectorConfig } from '../../../flow_deps/flow';

describe('getBindingIndex', () => {
    let defaultResponse: number, foundName: string;

    beforeEach(() => {
        defaultResponse = -1;
        foundName = 'acme/found';
    });

    describe('returns -1 when', () => {
        test('bindings are missing', () => {
            expect(getBindingIndex(undefined, '')).toBe(defaultResponse);
            expect(getBindingIndex(null, '')).toBe(defaultResponse);
        });

        test('bindings are empty', () => {
            expect(getBindingIndex([], '')).toBe(defaultResponse);
            expect(getBindingIndex([null], '')).toBe(defaultResponse);
            expect(getBindingIndex([undefined], '')).toBe(defaultResponse);
            expect(getBindingIndex([{}], '')).toBe(defaultResponse);
        });

        test('bindings do not contain collection', () => {
            expect(getBindingIndex(['a', 'b', 'c'], foundName)).toBe(-1);
        });
    });

    describe('returns index when ', () => {
        test('collection name is found', () => {
            // Bindings that are listed as strings
            expect(getBindingIndex([foundName, 'b', 'c'], foundName)).toBe(0);
            expect(getBindingIndex(['a', foundName, 'c'], foundName)).toBe(1);
            expect(getBindingIndex(['a', 'b', foundName], foundName)).toBe(2);
        });

        test('collection name is in a property of `target` or `source`', () => {
            expect(
                getBindingIndex(
                    [{ source: 'a' }, { source: foundName }, { source: 'c' }],
                    foundName
                )
            ).toBe(1);

            expect(
                getBindingIndex(
                    [{ target: 'a' }, { target: foundName }, { target: 'c' }],
                    foundName
                )
            ).toBe(1);
        });

        test('collection name is in a nested property of `name`', () => {
            expect(
                getBindingIndex(
                    [{ name: 'a' }, { name: foundName }, { name: 'c' }],
                    foundName
                )
            ).toBe(1);

            expect(
                getBindingIndex(
                    [
                        { source: { name: 'a' } },
                        { source: { name: foundName } },
                        { source: { name: 'c' } },
                    ],
                    foundName
                )
            ).toBe(1);

            expect(
                getBindingIndex(
                    [
                        { target: { name: 'a' } },
                        { target: { name: foundName } },
                        { target: { name: 'c' } },
                    ],
                    foundName
                )
            ).toBe(1);
        });
    });
});

describe('generateTaskSpec', () => {
    let mockedResourceConfig: ResourceConfig,
        connectorConfig: ConnectorConfig,
        resourceConfigs: ResourceConfigDictionary | null,
        existingTaskData: DraftSpecsExtQuery_ByCatalogName | null,
        sourceCapture: string | null,
        fullSource: FullSourceDictionary | null;

    beforeEach(() => {
        mockedResourceConfig = {
            errors: [],
            data: {
                fiz: 'resource',
            },
            disable: false,
            previouslyDisabled: false,
        };

        connectorConfig = generateMockConnectorConfig();
        resourceConfigs = null;
        existingTaskData = null;
        sourceCapture = null;
        fullSource = null;
    });

    describe('when no existing data or binding data', () => {
        beforeEach(() => {
            existingTaskData = null;
            resourceConfigs = {};
        });

        test('bindings will be empty', () => {
            const response = generateTaskSpec(
                'capture',
                connectorConfig,
                {},
                existingTaskData,
                sourceCapture,
                fullSource
            );
            expect(response.bindings).toStrictEqual([]);
        });
    });

    describe('when existing data but no binding data', () => {
        beforeEach(() => {
            connectorConfig = generateMockConnectorConfig();
            existingTaskData = {
                catalog_name: 'mock/test/capture-postgres',
                draft_id: '00:00:00:00:00:00:00:00',
                expect_pub_id: '00:00:00:00:00:00:00:00',
                spec_type: 'capture',
                spec: {
                    bindings: [
                        {
                            target: 'mock/binding/one',
                            resource: {
                                table: 'table_name',
                                delta_updates: false,
                            },
                        },
                        {
                            target: 'mock/binding/two',
                            resource: {
                                table: 'table_name_two',
                                delta_updates: false,
                            },
                            backfill: 2,
                        },
                    ],
                    endpoint: {
                        connector: connectorConfig,
                    },
                },
            };
        });

        test('existing bindings will be cleaned out as they assume they were deleted', () => {
            const response = generateTaskSpec(
                'capture',
                connectorConfig,
                {},
                existingTaskData,
                sourceCapture,
                fullSource
            );
            expect(response.bindings).toStrictEqual([]);
        });
    });

    describe('when existing data and new data', () => {
        beforeEach(() => {
            connectorConfig = generateMockConnectorConfig();
            existingTaskData = {
                catalog_name: 'mock/test/capture-postgres',
                draft_id: '00:00:00:00:00:00:00:00',
                expect_pub_id: '00:00:00:00:00:00:00:00',
                spec_type: 'capture',
                spec: {
                    bindings: [
                        {
                            target: 'mock/binding/one',
                            resource: {
                                table: 'table_name',
                                delta_updates: false,
                            },
                        },
                        {
                            target: 'mock/binding/two',
                            resource: {
                                table: 'table_name_two',
                                delta_updates: false,
                            },
                            backfill: 2,
                        },
                    ],
                    endpoint: {
                        connector: connectorConfig,
                    },
                },
            };
        });

        test('existing bindings will have their config overwritten and new ones added', () => {
            const response = generateTaskSpec(
                'capture',
                connectorConfig,
                {
                    'mock/binding/one': mockedResourceConfig,
                    'mock/binding/two': mockedResourceConfig,
                    'mock/binding/three': mockedResourceConfig,
                },
                existingTaskData,
                sourceCapture,
                fullSource
            );
            expect(response.bindings).toStrictEqual([
                {
                    resource: {
                        fiz: 'resource',
                    },
                    target: 'mock/binding/one',
                },
                {
                    backfill: 2,
                    resource: {
                        fiz: 'resource',
                    },
                    target: 'mock/binding/two',
                },
                {
                    resource: {
                        fiz: 'resource',
                    },
                    target: 'mock/binding/three',
                },
            ]);
        });

        describe('duplicates in existing bindings', () => {
            beforeEach(() => {
                if (!existingTaskData) {
                    return console.log('missing data');
                }

                existingTaskData.spec.bindings.push(
                    existingTaskData.spec.bindings[0]
                );
                existingTaskData.spec.bindings.push(
                    existingTaskData.spec.bindings[0]
                );
            });

            test('will update all copies in the bindings', () => {
                const response = generateTaskSpec(
                    'capture',
                    connectorConfig,
                    {
                        'mock/binding/one': {
                            ...mockedResourceConfig,
                            data: {
                                bar: 'updated',
                            },
                        },
                        'mock/binding/two': mockedResourceConfig,
                    },
                    existingTaskData,
                    sourceCapture,
                    fullSource
                );

                expect(response.bindings).toMatchSnapshot();
            });
        });
    });

    describe('entity type can alter what is returned', () => {
        beforeEach(() => {
            resourceConfigs = {
                first: mockedResourceConfig,
                second: mockedResourceConfig,
            };
        });

        test('captures use `target` and materializations use `source` as the binding name', () => {
            expect(
                generateTaskSpec(
                    'capture',
                    connectorConfig,
                    resourceConfigs,
                    existingTaskData,
                    sourceCapture,
                    fullSource
                ).bindings.map(({ target }: any) => target)
            ).toStrictEqual(['first', 'second']);

            expect(
                generateTaskSpec(
                    'materialization',
                    connectorConfig,
                    resourceConfigs,
                    existingTaskData,
                    sourceCapture,
                    fullSource
                ).bindings.map(({ source }: any) => source)
            ).toStrictEqual(['first', 'second']);
        });

        test('only materializations can return the `sourceCapture` property', () => {
            sourceCapture = 'mock/source/capture';

            expect(
                generateTaskSpec(
                    'capture',
                    connectorConfig,
                    resourceConfigs,
                    existingTaskData,
                    sourceCapture,
                    fullSource
                ).sourceCapture
            ).toBeUndefined();

            expect(
                generateTaskSpec(
                    'materialization',
                    connectorConfig,
                    resourceConfigs,
                    existingTaskData,
                    sourceCapture,
                    fullSource
                ).sourceCapture
            ).toStrictEqual(sourceCapture);
        });
    });
});
