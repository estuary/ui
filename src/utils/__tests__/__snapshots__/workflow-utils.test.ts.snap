// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateTaskSpec > when existing data and new data > duplicates in existing bindings > when the resource config contains a single instance of a duplicated binding, update one binding instance and remove others 1`] = `
[
  {
    "backfill": 2,
    "resource": {
      "fiz": "resource",
    },
    "target": "mock/binding/two",
  },
  {
    "resource": {
      "bar": "updated",
    },
    "target": "mock/binding/one",
  },
]
`;
